src/NHibernate/Type/ManyToOneType.cs

```diff
@@ -224,5 +204,31 @@ public override bool[] ToColumnNullness(object value, IMapping mapping)
 			}
 			return result;
 		}
+
+		private bool IsDirtyManyToOne(object old, object current, bool[] checkable, ISessionImplementor session)
+		{
+			if (IsSame(old, current))
+			{
+				return false;
+			}
+
+			if (old == null || current == null)
+			{
+				return true;
+			}
+
+			if (ForeignKeys.IsTransientFast(GetAssociatedEntityName(), current, session) == true)
```

Comment:
Need to use `IsTransientSlow`, otherwise the bug will still be there for entities having an assigned id already assigned or having an `unsaved-value` set as `undefined`. `IsTransientSlow` hits the db only if `IsTransientFast` was not able of answering, which occurs in aforementioned cases.